pkg/cri/server/container_events.go

```diff
@@ -21,13 +21,40 @@ import (
 )
 
 func (c *criService) GetContainerEvents(r *runtime.GetEventsRequest, s runtime.RuntimeService_GetContainerEventsServer) error {
-	// TODO (https://github.com/containerd/containerd/issues/7318):
-	// replace with a real implementation that broadcasts containerEventsChan
-	// to all subscribers.
-	for event := range c.containerEventsChan {
-		if err := s.Send(&event); err != nil {
-			return err
-		}
+	errCh := make(chan error, 1)
+
+	c.containerEventsClients.Store(s, errCh)
+
+	err := <-errCh
+
+	c.containerEventsClients.Delete(s)
+
+	return err
+}
+
+func (c *criService) broadcastEvents() {
+	for containerEvent := range c.containerEventsChan {
```

Comment:
With this implementation a single slow client will block other clients from receiving updates.
Also each new client slows down delivery since these are being sent sequentially.

Reply:
@cpuguy83 Thank you for highlighting the concerns. I'm considering wrapping the range function in a goroutine to address the blocking issue. Would it be reasonable to limit the maximum routines running concurrently? If so, what do you think would be a good limit here?

Reply:
Is there any reason subscribers need to attach to the same event stream over creating new streams for each subscriber?

Reply:
The subscribers need to receive the same events. We can consider doing a fan out from a different place. For example here https://github.com/containerd/containerd/blob/7deb68fbf4b550d089488840d5ba15c7e345b6db/pkg/cri/server/helpers.go#L446
Is this what you have in mind?